### PR TITLE
Feature/custom allocator distributed closest point

### PR DESCRIPTION
**Pull request by gunney1**
# Support custom allocator in `DistributedClosestPoint` class and its example

- This PR is a feature
- It does the following:
  - Adds `setAllocatorId` to `DistributedClosestPoint`.
  - Chooses host or device memory allocation in the `quest_distributed_distance_query_example` according to the execution policy.
  - Tests the use of device memory by `DistributedClosestPoint`.

This PR resolves issue https://github.com/LLNL/axom/issues/925

### CODE REVIEW

File: src/axom/quest/DistributedClosestPoint.hpp [hunk 1]
```diff
@@ -369,7 +369,7 @@ class DistributedClosestPointImpl
     , m_rank(-1)
     , m_nranks(-1)
   {
-    setAllocatorID();
+    setDefaultAllocatorID();
 
     setMpiCommunicator(MPI_COMM_WORLD);
```
Comment (rhornung67) [inline src/axom/quest/DistributedClosestPoint.hpp:374]: Should this be changed to allow user to pass their own MPI communicator? I foresee cases when codes are using this in 'multiprob' execution mode.

Comment (kennyweiss) [reply src/axom/quest/DistributedClosestPoint.hpp:374]: @gunney1 already added that in #903 
and it resolved the issue where we couldn't run more than one test of this example at the same time.

File: src/axom/quest/DistributedClosestPoint.hpp [hunk 2]
```diff
@@ -396,6 +396,14 @@ class DistributedClosestPointImpl
     m_sqDistanceThreshold = sqThreshold;
   }
 
+  /// Sets the allocator ID to the default associated with the execution policy
+  void setAllocatorID(int allocatorID)
+  {
+    SLIC_ASSERT(allocatorID != axom::INVALID_ALLOCATOR_ID);
```
Comment (kennyweiss) [inline src/axom/quest/DistributedClosestPoint.hpp:402]: Thought: Should we allow the user to set it to `axom::INVALID_ALLOCATOR_ID`? e.g. to clear it to the default?

**Review by kennyweiss — APPROVED** [review]
Thanks @gunney1 -- this looks great!

File: src/axom/quest/DistributedClosestPoint.hpp [hunk 3]
```diff
@@ -1301,6 +1309,14 @@ class DistributedClosestPoint
     return false;
   }
 
+  /// Sets the allocator ID to the default associated with the execution policy
```
Comment (agcapps) [inline src/axom/quest/DistributedClosestPoint.hpp:1312]: Please update Doxygen

File: src/axom/quest/examples/quest_distributed_distance_query_example.cpp [hunk 4]
```diff
@@ -1026,6 +1026,28 @@ int main(int argc, char** argv)
...
+  #elif defined(UMPIRE_ENABLE_UM)
+    "UM"
+  #elif defined(UMPIRE_ENABLE_PINNED)
+    "PINNED"
+  #else
+    "HOST"
+  #endif
+    ;
+  auto& rm = umpire::ResourceManager::getInstance();
+  umpire::Allocator ummpireAllocator = rm.getAllocator(umpireResourceName);
```
Comment (agcapps) [inline src/axom/quest/examples/quest_distributed_distance_query_example.cpp:1048]: please adjust spelling of `ummpireAllocator`

**Review by agcapps — APPROVED** [review]
Thanks, @gunney1 !

File: src/axom/quest/DistributedClosestPoint.hpp (hunk 2 above)
Comment (gunney1) [reply src/axom/quest/DistributedClosestPoint.hpp:402]: Interesting idea.  How about if we make `setDefaultAllocatorID` public, to do the same thing explicitly?

File: src/axom/quest/examples/quest_distributed_distance_query_example.cpp (hunk 4 above)
Comment (gunney1) [reply src/axom/quest/examples/quest_distributed_distance_query_example.cpp:1048]: OK.  Thanks for catching.

File: src/axom/quest/DistributedClosestPoint.hpp (hunk 3 above)
Comment (gunney1) [reply src/axom/quest/DistributedClosestPoint.hpp:1312]: Will do.

**Pull request merged**